src/index.ts

```diff
@@ -35,73 +36,132 @@ let traceAgent: StackdriverTracer;
-    config.rootSpanNameOverride = () => spanName;
-  } else if (typeof config.rootSpanNameOverride !== 'function') {
-    config.rootSpanNameOverride = (name: string) => name;
-  }
-
-  // If the CLS mechanism is set to auto-determined, decide now what it should
-  // be.
-  const ahAvailable = semver.satisfies(process.version, '>=8');
-  if (config.clsMechanism === 'auto') {
-    config.clsMechanism = ahAvailable ? 'async-hooks' : 'async-listener';
-  }
-
-  return config;
+  const projectConfig: (typeof defaultConfig)&Forceable<Config> =
+      extend(true, {}, defaultConfig, envSetConfig, userConfig);
+  const forceNew = userConfig[FORCE_NEW];
+
+  return {
+    [FORCE_NEW]: forceNew,
+    enabled: projectConfig.enabled,
+    logLevel: lastOf(
+        projectConfig.logLevel, Number(process.env.GCLOUD_TRACE_LOGLEVEL)),
+    clsConfig: {
+      [FORCE_NEW]: forceNew,
+      mechanism: transform(
+          projectConfig.clsMechanism,
+          (clsMechanism):
+              TraceCLSMechanism => {
+                // If the CLS mechanism is set to auto-determined, decide now
+                // what it should be.
+                const ahAvailable = semver.satisfies(process.version, '>=8');
+                if (clsMechanism === 'auto') {
+                  return ahAvailable ? TraceCLSMechanism.ASYNC_HOOKS :
+                                       TraceCLSMechanism.ASYNC_LISTENER;
+                }
+                return clsMechanism as TraceCLSMechanism;
+              })
+    },
+    writerConfig: {
+      [FORCE_NEW]: forceNew,
+      projectId: lastOf<string|undefined>(
+          projectConfig.projectId, process.env.GCLOUD_PROJECT),
+      onUncaughtException: projectConfig.onUncaughtException,
+      bufferSize: projectConfig.bufferSize,
+      flushDelaySeconds: projectConfig.flushDelaySeconds,
+      stackTraceLimit: projectConfig.stackTraceLimit,
+      maximumLabelValueSize: transform(
+          projectConfig.maximumLabelValueSize,
+          (maximumLabelValueSize) => {
+            // Enforce the upper limit for the label value size.
+            if (maximumLabelValueSize >
+                Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT) {
+              return Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT;
+            }
+            return maximumLabelValueSize;
+          }),
+      serviceContext: {
+        service: lastOf<string|undefined>(
+            projectConfig.serviceContext.service, process.env.GAE_MODULE_NAME,
+            process.env.GAE_SERVICE),
+        version: lastOf<string|undefined>(
+            projectConfig.serviceContext.version,
+            process.env.GAE_MODULE_VERSION, process.env.GAE_VERSION),
+        minorVersion: lastOf<string|undefined>(
+            projectConfig.serviceContext.minorVersion,
+            process.env.GAE_MINOR_VERSION)
+      }
+    },
+    pluginLoaderConfig: {
+      [FORCE_NEW]: forceNew,
+      plugins: Object.assign({}, projectConfig.plugins),
```

Comment:
style nit: `plugins: {... projectConfig.plugins)`

Reply:
IIRC this is not available in Node 6... which we haven't officially ripped out of CI yet.

Reply:
it is available in TypeScript tho

Reply:
We have to target es2015 to preserve generators, so it's not available even in TypeScript, unfortunately.

Reply:
Ok, that being said I'll probably just drop node6 CI in a different PR.

Reply:
Done